src/caffeine/caffeine_assert_s.F90

```diff
@@ -0,0 +1,36 @@
+! Copyright (c), The Regents of the University of California
+! Terms of use are as specified in LICENSE.txt
+submodule(prif:prif_private_s) caffeine_assert_s
+  implicit none
+
+#if CAF_ASSERTIONS || !defined(CAF_ASSERTIONS)
+  logical, parameter :: assertions_=.true.
+#else
+  logical, parameter :: assertions_=.false.
+#endif
+
+   !! Disable assertions with
+   !! fpm test --flag "-DCAF_ASSERTIONS=0"
+
+contains
+
+  module procedure assert
+    character(len=:), allocatable :: tail
+
+    if (assertions_) then 
+      if (.not. present(diagnostics)) then
+        tail = "."
+      else
+        tail = " with diagnostics " 
+        select type(diagnostics)
+          type is(character(len=*))
+            tail = tail // diagnostics
+          class default
+            tail = tail // "of unsupported type."
+        end select
```

Comment:
The block of code highlighted above nominally executes on every `assert()` call when assertions are enabled and `diagnostic` is present. However the `tail` it allocates and populates is only used if the assertion fails. 

This should not be a correctness problem (assuming `diagnostic` is not longer than the available memory), but it's somewhat wasteful of performance. Performance in assertion-enabled mode is not a primary concern, but IIUC it's just as easy to relocate this code into the `.not. assertion` branch below so it only executes when we actually need the result.

Reply:
@bonachea Have updated the PR based on your comments.
@rouson Since you handed off the finishing up of this PR to me, I have addressed Dan's above review comments, but feel free to check out the changes I made, as it changed restructured the code you wrote a bit.